clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -892,34 +903,118 @@ public ListOffsetResult() {
-                    partition, position, node);
+                log.debug("Added {} fetch request for partition {} at position {} to node {}", isolationLevel,
+                    partition, position, leaderAndEpoch.leader);
             }
         }
+
+        if (!partitionsToValidate.isEmpty()) {
+            // Send off the OffsetsForLeader requests asynchronously
+            Map<Node, Map<TopicPartition, OffsetsForLeaderEpochRequest.PartitionData>> regrouped =
+                    regroupPartitionMapByNode(partitionsToValidate);
+
+            regrouped.forEach((node, dataMap) -> {
+                OffsetsForLeaderEpochRequest.Builder builder = new OffsetsForLeaderEpochRequest.Builder(
+                        ApiKeys.OFFSET_FOR_LEADER_EPOCH.latestVersion(), dataMap);
+                sendOffsetForLeaderEpochRequest(node, builder);
+            });
+        }
+
         Map<Node, FetchSessionHandler.FetchRequestData> reqs = new LinkedHashMap<>();
         for (Map.Entry<Node, FetchSessionHandler.Builder> entry : fetchable.entrySet()) {
             reqs.put(entry.getKey(), entry.getValue().build());
         }
         return reqs;
     }
 
+    private <T> Map<Node, Map<TopicPartition, T>> regroupPartitionMapByNode(Map<TopicPartition, T> partitionMap) {
+        return partitionMap.entrySet()
+                .stream()
+                .collect(Collectors.groupingBy(entry -> metadata.fetch().leaderFor(entry.getKey()),
+                        Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue)));
+    }
+
+    private RequestFuture<OffsetsForLeaderEpochResponse> sendOffsetForLeaderEpochRequest(
+            Node node, OffsetsForLeaderEpochRequest.Builder builder) {
+
+        return client.send(node, builder).compose(new RequestFutureAdapter<ClientResponse, OffsetsForLeaderEpochResponse>() {
+            @Override
+            public void onSuccess(ClientResponse response, RequestFuture<OffsetsForLeaderEpochResponse> future) {
+                OffsetsForLeaderEpochResponse ofler = (OffsetsForLeaderEpochResponse) response.responseBody();
+                log.trace("Received OffsetsForLeaderEpochResponse {} from broker {}", ofler, node);
+
+                Map<TopicPartition, OffsetAndMetadata> truncationWithoutResetPolicy = new HashMap<>();
+                ofler.responses().forEach((respTopicPartition, respEndOffset) -> {
+                    // For each OffsetsForLeader response, check for truncation. This occurs iff the epoch has changed
+                    // and the end offset is lower than our current offset
+                    ConsumerMetadata.LeaderAndEpoch currentLeaderAndEpoch = metadata.leaderAndEpoch(respTopicPartition);
+                    ConsumerMetadata.LeaderAndEpoch leaderAndEpochForEndOffset = new ConsumerMetadata.LeaderAndEpoch(
+                            Node.noNode(), Optional.of(respEndOffset.leaderEpoch()));
+
+                    if (currentLeaderAndEpoch.isObsoletedBy(leaderAndEpochForEndOffset)) {
```

Comment:
Again my fault, but we can probably come up with a better name for `isObsoletedBy`. Doesn't read very fluently.

Can you explain the purpose of this check? I think the end offset epoch returned in the response will typically be less than the current epoch. When we receive the response, I think we need to check:

1. Does the partition still need validation for the epoch that was requested? This seems a little tricky because if the partition leadership changes quickly, we may need to validate the new epoch before we have finished validating a previous one.
2. Is the end offset returned from the leader less than the fetch offset? Again we may need to check again whether the fetch offset has changed while this request was inflight.

Initially I was thinking that we needed to do some checking of the epoch returned in the response as well, but now I think the offset may be the only thing. The OffsetForLeaderEpoch API will return the end offset of the largest epoch which is less than or equal to the requested epoch. The epoch returned is only useful for brokers . However it is still helpful to include it in the LogTruncation error that we raise below.

Reply:
So should we just replace the `isObsoletedBy` check with something like `subscriptions.awaitingValidation(respTopicPartition)`?